### PR TITLE
Avoid executing no-op queries.

### PR DESCRIPTION
**Pull request by clokep**
Return early for `simple_insert_many_txn`, `simple_upsert_many_txn`, and `simple_update_many_txn` if the iterable to update is empty.

Note that we already do this for `simple_select_many_txn` and `simple_delete_many_txn`.

This avoids hitting the database with a no-op query, the downside is that we can't pass generators into this function.

### CODE REVIEW

File: synapse/storage/database.py [hunk 1]
```diff
@@ -1483,6 +1486,9 @@ def simple_upsert_many_txn(
             value_values: A list of each row's value column values.
                 Ignored if value_names is empty.
         """
+        if not value_values:
```
Comment (erikjohnston) [inline synapse/storage/database.py:1489]: I think this needs to be `not key_values and not value_values`

Comment (clokep) [reply synapse/storage/database.py]: Hah, I just changed that. 👍

Comment (erikjohnston) [reply synapse/storage/database.py]: haha, great minds

File: synapse/storage/database.py [hunk 2]
```diff
@@ -2059,6 +2065,8 @@ def simple_update_many_txn(
             raise ValueError(
                 f"{len(key_values)} key rows and {len(value_values)} value rows: should be the same number."
             )
+        if not key_values:
```
Comment (erikjohnston) [inline synapse/storage/database.py:2068]: I think we should also check the `value_values`? I *think* it's probably fine, but we do specifically handle the case of an empty `key_names` below.

File: synapse/storage/database.py [hunk 3]
```diff
@@ -1483,6 +1486,9 @@ def simple_upsert_many_txn(
             value_values: A list of each row's value column values.
                 Ignored if value_names is empty.
         """
+        if not key_values:
```
Comment (erikjohnston) [inline synapse/storage/database.py:1489]:
```suggestion
        if not key_values and not value_values:
```

I think `key_names` and `key_values` have to be empty to make sense, but still

Comment (clokep) [reply synapse/storage/database.py:1489]: I'm not sure what you mean; I guess this could be used as:

```python
db_pool.simple_upsert_many_txn(
    txn,
    table="foo",
    key_names: (),
    key_values: [(), ()],
    value_names: ("a",),
    value_values: [(1,), (2,)],
)
```

tl;dr I think `key_values` and `value_values` need to be the same length?

Comment (erikjohnston) [reply synapse/storage/database.py:1489]: Ah, OK fair.

File: synapse/storage/database.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/database.py:2068]: Ah, didn't realise they needed to be the same length

**Review by erikjohnston — APPROVED** [review]

File: synapse/storage/database.py (hunk 3 above)
Comment (clokep) [reply synapse/storage/database.py:1489]: It looks like you *can* provide only `key_values` and not `value_values` and we'll automatically fill those in, so I *think* this is the correct change to make.

**Comment by clokep** [comment]
Once #16596 is merged, I'll add tests for these changes to ensure nothing is breaking.

**Comment by clokep** [comment]
I added some additional tests (based on #16596) and some more comments. 👍
